plugins/dfu/dfu-target-avr.c

```diff
@@ -228,6 +291,126 @@ dfu_target_avr_read_memory (DfuTarget *target,
 	return TRUE;
 }
 
+/**
+ * dfu_target_avr_read_command:
+ * @target: a #DfuTarget
+ * @memory_unit: a unit, e.g. %DFU_AVR32_MEMORY_UNIT_FLASH
+ * @cancellable: a #GCancellable, or %NULL
+ * @error: a #GError, or %NULL
+ *
+ * Performs a read operation on the device.
+ *
+ * Return value: %TRUE for success
+ **/
+static gboolean
+dfu_target_avr_read_command (DfuTarget *target, guint8 page, guint8 addr,
+			     GCancellable *cancellable, GError **error)
+{
+	g_autoptr(GBytes) data_in = NULL;
+	guint8 buf[3];
+
+	/* format buffer */
+	buf[0] = DFU_AVR_CMD_READ_COMMAND;
+	buf[1] = page;
+	buf[2] = addr;
+	data_in = g_bytes_new_static (buf, sizeof(buf));
+	g_debug ("read command page:0x%02x addr:0x%02x", (guint) page, (guint) addr);
+	if (!dfu_target_download_chunk (target, 0, data_in, cancellable, error)) {
+		g_prefix_error (error, "cannot read command page: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * dfu_target_avr32_get_chip_signature:
+ * @target: a #DfuTarget
+ * @cancellable: a #GCancellable, or %NULL
+ * @error: a #GError, or %NULL
+ *
+ * Gets the chip signature for the AVR32 device.
+ *
+ * Return value: a 4-byte %GBytes object for success, else %NULL
+ **/
+static GBytes *
+dfu_target_avr32_get_chip_signature (DfuTarget *target,
+				     GCancellable *cancellable,
+				     GError **error)
+{
+	/* select unit, and request 4 bytes */
+	if (!dfu_target_avr_select_memory_unit (target,
+						DFU_AVR32_MEMORY_UNIT_SIGNATURE,
+						cancellable, error))
+		return NULL;
+	if (!dfu_target_avr32_select_memory_page (target, 0x00, cancellable, error))
+		return NULL;
+	if (!dfu_target_avr_read_memory (target, 0x00, 0x03, cancellable, error))
+		return NULL;
+
+	/* get data back */
+	return dfu_target_upload_chunk (target, 0x00, 0, cancellable, error);
+}
+
+/**
+ * dfu_target_avr_get_chip_signature:
+ * @target: a #DfuTarget
+ * @cancellable: a #GCancellable, or %NULL
+ * @error: a #GError, or %NULL
+ *
+ * Gets the chip signature for the AVR device.
+ *
+ * Return value: a 4-byte %GBytes object for success, else %NULL
+ **/
+static GBytes *
+dfu_target_avr_get_chip_signature (DfuTarget *target,
+				   GCancellable *cancellable,
+				   GError **error)
+{
+	struct {
+		guint8 page;
+		guint addr;
+	} signature_locations[] = {
+		{ 0x01, 0x30 },
+		{ 0x01, 0x31 },
+		{ 0x01, 0x60 },
+		{ 0x01, 0x61 },
+		{ 0xff, 0xff }
+	};
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* we have to request this one byte at a time */
+	chunks = g_ptr_array_new_with_free_func ((GDestroyNotify) g_bytes_unref);
```

Comment:
This isn't a double free by using g_autoptr for chunks and setting a free function?

Reply:
Oh nevermind, that's for freeing the individual elements.  Carry on :)

Reply:
Ya, it's a bit unusual to have an array of 1-byte GBytes that's then converted to a 4-byte GBytes -- but this is the wonderful world of hardware hacking :)